guide.md

```diff
@@ -0,0 +1,324 @@
+    srcs = ["hello-world.cc"],
+)
+```
+
+and `src/hello-world.cc` is simply
+
+```cpp
+#include <iostream>
+
+int main(int argc, char** argv) {
+  std::cout << "Hello world!" << std::endl;
+  return 0;
+}
+```
+
+Verify that this builds:
+
+```
+> bazel run //src:hello-world
+INFO: Analyzed target //src:hello-world (0 packages loaded, 0 targets configured).
+INFO: Found 1 target...
+Target //src:hello-world up-to-date:
+  bazel-bin/src/hello-world
+INFO: Elapsed time: 0.063s, Critical Path: 0.00s
+INFO: 1 process: 1 internal.
+INFO: Build completed successfully, 1 total action
+INFO: Build completed successfully, 1 total action
+Hello world!
+```
+
+# The lack of hermeticity
+
+Bazel is valued for its hermetic and reproducible builds, but this example
+already shows that that isn't completely accurate: we haven't told Bazel what
+C++ compiler to use or where to find it and all the auxiliary tools it relies
+on (linker, assembler, and the like). It simply looks for appropriately-named
+executables in the PATH and sets up a C/C++ compiler toolchain called
+`@local_config_cc` that "should work". 
+
+This is problematic, and not just because of academic concerns regarding the
+precise meaning of "hermeticity": build products output by different versions
+of the same compiler can vary greatly, let alone between different compilers.
+Bazel, by default, chooses GCC on a Linux machine and Clang/LLVM on a macOS
+machine, which can mean that the same bit of hot code, say, vectorizes
+correctly on one platform but not on the other due to some difference between
+the two compilers, resulting in wildly different performance characteristics.
+
+There are Bazel-native ways to mitigate these issues, but they fall short of true
+hermeticity: for instance, `--incompatible_strict_action_env` runs actions in
+an environment that does not inherit a `PATH` from the shell environment, but
+instead [sets it to][path-link] the static value `/bin:/usr/bin:/usr/local/bin`.
+This means that as long as your compiler tools are in one of those paths, which
+they almost certainly are on \*nix systems, nothing changes.
+
+# How rules_nixpkgs solves the issue
+
+There are two classes of workaround: one is to give up trying to make Bazel not
+depend on the global environment, and instead to control that global
+environment ourselves so that the implicit dependencies end up being the same
+every time. To do so, we run all builds in sandboxes or VMs that all share the
+exact same configuration, for instance, in identical Docker containers.
+
+The other solution is to eliminate the dependency on the global environment by
+defining your own compiler toolchain, providing paths to executable binaries that you
+provision and version yourself. Bazel can then be instructed to use that
+instead of the default `@local_config_cc`. This can be done by checking all the
+relevant executables (`cc`, `ld`, `as`, ...) into your repository yourself, as
+is done within Google for instance. 
```

Comment:
```suggestion
relevant executables (`cc`, `ld`, `as`, ...) into your repository yourself. 
```
Not sure if they've documented that properly and publically anywhere. I've been told that's the case, but that's just hearsay.